User returns a book

margaretoti opened the pull request:
**UPDATED:** Users can now return books that they checked out via the new profile page. A return link will only appear near each checked out book. Once a book is returned, it will be available for other users to check out. 

<img width="465" alt="screen shot 2017-04-14 at 12 42 38 pm" src="https://cloud.githubusercontent.com/assets/9501674/25049444/df74923c-210f-11e7-9e21-2d3e96601d67.png">

### Profile Page
<img width="462" alt="screen shot 2017-04-14 at 12 42 30 pm" src="https://cloud.githubusercontent.com/assets/9501674/25049451/e44201dc-210f-11e7-9825-0395bb53968d.png">

margaretoti committed:
User can return books to the library

User can return books to the library via the books index page. Only
books that the current user has checked out have a return link next to
them.

margaretoti committed:
Refactor tests and controller

margaretoti commented on app/assets/stylesheets/application.scss:
```diff
@@ -1,9 +1,9 @@
-
-@import "base/base";
-@import "refills/flashes";
+// @charset "utf-8";
```
I removed the default styles as I will worry about styling the application later.

margaretoti commented on spec/support/features/checkouts_helper.rb:
```diff
@@ -0,0 +1,13 @@
+    have_css('ul.books.available li', text: title)
+  end
+
+  def have_available_book(title)
```
def `have_available_book` and def `have_returned_book` do the same thing. However, the context of the test makes one more readable than the other. Is it okay to have two functions that do the same thing if readability is improved?

See [spec/features/user_checks_out_book_spec.rb](https://github.com/margaretoti/library-network/pull/9/files#diff-a8417d893f833dade4dbb49bf25eb35d) and [spec/features/user_returns_checked_out_book_spec.rb](https://github.com/margaretoti/library-network/pull/9/files#diff-b2009e0746566ff3a7cd288d60c76629)

derekprior replied:
I think that's ok. I'd perhaps consider using [`alias_method`](http://ruby-doc.org/core-2.4.1/Module.html#method-i-alias_method) rather than duplicating the definition. RSpec itself has many aliases for its matchers. On one hand it can be confusing, on the other hand sometimes being able to choose between them can really aid in readability/understandability

margaretoti commented on spec/features/user_returns_checked_out_book_spec.rb:
```diff
@@ -0,0 +1,24 @@
+  scenario "successfully" do
+    book_1 = create(:book)
+    patron_1 = create(:patron)
+    sign_in_and_check_out_book(patron_1)
```
def `sign_in_and_check_out_book` can probably also handle book and patron creation, except I wanted to make it clear that book creation is an not completed by the user. Additionally, creating a patron inside the function made it very difficult to tell that after sign-out a different user is signing in. Any advice you have on how to keep the functions small, but also keep the clear and readable would be great.

derekprior replied:
Yeah. I don't think I'd bother extracting that helper method at all. `sign_in_and_check_out` seem like two different steps that I'd be happy to see repeated as often as they need to be.

```ruby
visit books_path(as: user)
click_on "Check out"
```

seems quite readable to me.

derekprior commented on app/controllers/checkouts_controller.rb:
```diff
@@ -7,4 +7,12 @@ def create
+  def destroy
+    @book = Book.find(params[:book_id])
+    @checkout = Checkout.where(user: current_user, book: @book)
+    @checkout.destroy_all
```
I think you can accomplish this all in one line - or at least 1 statement:

```ruby
Checkout.
  where(user_id: current_user, book_id: params[:book_id]).
  destroy_all
```

I don't think you need to assign any instance variables here as we're not redering them in any views.

derekprior replied:
Actually, thinking about this some more -- does it really matter which user returned the book? If I think about the way my library works, I don't identify myself when returning books. Each book is uniquely identified and when someone (anyone) returns it, the checkout is removed/ended.

I wonder if we could use methods on the models to encapsulate the behavior we want here. A few options I could see:

```ruby
Book.find(params[:book_id]).return
Checkout.find(params[:book_id]).return
Return.process(book)
```

The first would feel a little weird because it would require that the `Book` model know how to handle, for example, deleting or updating a checkout instance.

The second seems okay. It's not exactly natural language to say "return this checkout". From an OO perspective it seems like the right object to receive some sort of message and handle this though.

The third option is a plain ruby object (not an ActiveRecord instance) that just knows how to process returns. It's probably premature to go to something like this right now, but it would be a good option if returns became a multi-step process at some point. Something to keep in mind.

In any case, we're encoding the knowledge that returning a book means destroying the checkout record in a method. If at some point we decide "no, we just want to have a `returned_at` column rather than deleting the record` we could update the model and not the controller or anywhere else where returns may happen.

The code you already have here is simple enough that all of those options may be overkill at this point, but I thought it at least instructive to consider it and at the very least keep it in mind as an option as we move forward if checkouts get more complicated, for instance.

What do you think?

margaretoti replied:
@derekprior, Thanks for this great explanation. You are right; you don't need to identify yourself. At a library, I just place my books in the return trolley and the unique record of that checkout it destroyed/archived once an admin scans it as a returned item. 

I like options 2 and 3. I admit 2 is slightly more familiar to me, but I am definitely open to trying option 3 as I have  less experience dealing with plain Ruby objects. 

I also agree with your comment down below. This UI serves as a way to get things working quickly. A dedicated view where users see the books they have checked out (and return them) is much better, and what I planned to build next. That way the list of unavailable books on the books `index` page does not have to worry about who can return which books. 

Is the best course of action to change this PR so that it uses a user profile page (or something similar) to handle book returns and viewing currently checked out books? That is what I lean.

derekprior commented on app/views/books/index.html.erb:
```diff
@@ -15,8 +15,11 @@
-      <%= book.title %>, 
+      <%= book.title %>,
       <%= book.author %>
+      <% if current_user.checkouts.where(book: book).present? %>
```
There's a couple things worth discussing here.

1. We typically want to avoid `where` and the like in views. They are far removed from the model. Using `where` means the view is now coupled to the ActiveRecord implementation of `Checkout`.

2. This introduces an N+1 query. That is, `@unavailable_books` performs one query to get a list of books. Then for the `N` books we have we're going to end up doing `N` SQL queries to check for the existence of a checkout of that book by this user. When there's a small number of books, this is inefficient, but perhaps not noticeable. If there are 10, 100, 1000 books then the performance hit of the database round trip becomes a problem.

The quick way to fix both of these and maintain this UI is to do something like: `@current_user_book_ids = current_user.checkouts.pluck(:book_id)` in the controller. This will do a single query to return an array of book ids the user has a checkout for. Then here in the view do something like `if @current_user_book_ids.include?(book_id)`

The consultant in me would probably ask, "Is what we're doing here really the best way for a user to to return their books?" I'd probably expect to have a page dedicated to viewing the books *I* have checked out and the ability to return them? What's here might be the quickest way to get something workable, but I suspect I'd be technically happier with the dedicated view if it were a possibility.

margaretoti committed:
Add basic user profile

User can now see his or her checkouts on the user profile page. It is on
this page that they will be able to return checked out books.

margaretoti committed:
User can return books from her profile

Books can no longer be returned from the book index page. When a user
returns a book from her profile, she is redirected back to her profile
page.

Next steps include: 1) Adding a link to the profile page on the books
index. 2) Adding a link to the books index to the profile page.

margaretoti committed:
Refactor user returns book test

derekprior commented on app/controllers/profiles_controller.rb:
```diff
@@ -0,0 +1,15 @@
+
+  def show
+    checkouts = current_user.checkouts
+    checked_out_book_ids = checkouts.pluck(:book_id)
```
You don't need to use this method (which I think I had suggested in the previous PR iteration). I had suggested it when you were displaying *all* of the checked out books and then trying to add a "return" link if the book was checked out by the current user. Since this controller action and view only shows books the user has checked out, you can do things more conventionally or directly.

You've already got all of the users checkout records with `current_user.checkouts`. You can access the book directly in the view from there when iterating like so:

```
<% @checkouts.each do |checkout| %>
  <li> <%= checkout.book.title %></li>
<% end %>
```

If you can update the code to do this, I think it will simplify this controller action some.

derekprior reviewed (commented):
Looking good. Just one last bit of deedback about the controller.

derekprior commented on app/models/profile.rb:
```diff
@@ -0,0 +1,2 @@
+class Profile < ApplicationRecord
```
Does this need to exist?

derekprior commented on app/views/profiles/show.html.erb:
```diff
@@ -0,0 +1,11 @@
+    <li>
+      <%= checkout.book.title %>,
+      <%= checkout.book.author %>
+      <%= link_to "Return", book_checkout_path(checkout.book, current_user), method: :delete %>
```
Should this be a translation?

margaretoti committed:
Add eager loading to checkouts
